File: CS/ChartsDemo.Wpf/ViewModels/RealtimeViewModel.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Windows.Threading;

namespace ChartsDemo {
    public class RealtimeViewModel {
        const int UpdateInterval = 40;

        readonly DataCollection dataSource = new DataCollection();
        readonly DispatcherTimer timer = new DispatcherTimer(DispatcherPriority.Normal);
        readonly Random random = new Random();
        double value1 = 10.0;
        double value2 = -10.0;
        bool inProcess = true;

        public int TimeInterval { get; set; }
        public virtual bool IsTimerEnabled { get; set; }
        public DataCollection DataSource { get { return dataSource; } }
        public virtual DateTime MinTime { get; protected set; }
        public virtual DateTime MaxTime { get; protected set; }

        public RealtimeViewModel() {
            IsTimerEnabled = true;
            TimeInterval = 10;
            timer.Interval = TimeSpan.FromMilliseconds(UpdateInterval);
            timer.Tick += (_d, _e) => OnTimerTick();
        }

        public void DisableProcess() {
            inProcess = timer.IsEnabled;
            IsTimerEnabled = false;
        }
        public void RestoreProcess() {
            IsTimerEnabled = inProcess;
        }
        public void ToggleIsTimerEnabled() {
            IsTimerEnabled = !IsTimerEnabled;
        }

        protected void OnIsTimerEnabledChanged() {
            timer.IsEnabled = IsTimerEnabled;
        }
        void OnTimerTick() {
            DateTime argument = DateTime.Now;
            DateTime minDate = argument.AddSeconds(-TimeInterval);
            IList<ProcessItem> itemsToInsert = new List<ProcessItem>();
            for(int i = 0; i < UpdateInterval; i++) {
                itemsToInsert.Add(new ProcessItem() { DateAndTime = argument, Process1 = value1, Process2 = value2 });
                argument = argument.AddMilliseconds(1);
                value1 = CalculateNextValue(value1);
                value2 = CalculateNextValue(value2);
            }
            dataSource.AddRange(itemsToInsert);
            dataSource.RemoveRangeAt(0, dataSource.TakeWhile(item => item.DateAndTime < minDate).Count());

            MinTime = minDate;
            MaxTime = argument;
        }
        double CalculateNextValue(double value) {
            return value + (random.NextDouble() * 10.0 - 5.0);
        }
    }

    public struct ProcessItem {
        public DateTime DateAndTime { get; set; }
        public double Process1 { get; set; }
        public double Process2 { get; set; }
    }

    public class DataCollection : ObservableCollection<ProcessItem> {
        public void AddRange(IList<ProcessItem> items) {
            foreach(ProcessItem item in items)
                Items.Add(item);
            OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, (IList)items, Items.Count - items.Count));
        }
        public void RemoveRangeAt(int startingIndex, int count) {
            var removedItems = new List<ProcessItem>(count);
            for(int i = 0; i < count; i++) {
                removedItems.Add(Items[startingIndex]);
                Items.RemoveAt(startingIndex);
            }
            if(count > 0)
                OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, (IList)removedItems, startingIndex));
        }
    }
}

File: VB/ChartsDemo.Wpf/ViewModels/RealtimeViewModel.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Collections.Specialized
Imports System.Linq
Imports System.Text
Imports System.Windows.Threading

Namespace ChartsDemo
    Public Class RealtimeViewModel
        Private Const UpdateInterval As Integer = 40


        Private ReadOnly dataSource_Renamed As New DataCollection()
        Private ReadOnly timer As New DispatcherTimer()
        Private ReadOnly random As New Random()
        Private value1 As Double = 10.0
        Private value2 As Double = -10.0
        Private inProcess As Boolean = True

        Public Property TimeInterval() As Integer
        Public Overridable Property IsTimerEnabled() As Boolean
        Public ReadOnly Property DataSource() As DataCollection
            Get
                Return dataSource_Renamed
            End Get
        End Property
        Public Overridable Property MinTime() As Date
        Public Overridable Property MaxTime() As Date

        Public Sub New()
            IsTimerEnabled = True
            TimeInterval = 10
            timer.Interval = TimeSpan.FromMilliseconds(UpdateInterval)
            AddHandler timer.Tick, Sub(_d, _e) OnTimerTick()
        End Sub

        Public Sub DisableProcess()
            inProcess = timer.IsEnabled
            IsTimerEnabled = False
        End Sub
        Public Sub RestoreProcess()
            IsTimerEnabled = inProcess
        End Sub
        Public Sub ToggleIsTimerEnabled()
            IsTimerEnabled = Not IsTimerEnabled
        End Sub

        Protected Sub OnIsTimerEnabledChanged()
            timer.IsEnabled = IsTimerEnabled
        End Sub
        Private Sub OnTimerTick()
            Dim argument As Date = Date.Now
            Dim minDate As Date = argument.AddSeconds(-TimeInterval)
            Dim itemsToInsert As IList(Of ProcessItem) = New List(Of ProcessItem)()
            For i As Integer = 0 To UpdateInterval - 1
                itemsToInsert.Add(New ProcessItem() With {.DateAndTime = argument, .Process1 = value1, .Process2 = value2})
                argument = argument.AddMilliseconds(1)
                value1 = CalculateNextValue(value1)
                value2 = CalculateNextValue(value2)
            Next i
            dataSource_Renamed.AddRange(itemsToInsert)
            dataSource_Renamed.RemoveRangeAt(0, dataSource_Renamed.TakeWhile(Function(item) item.DateAndTime < minDate).Count())

            MinTime = minDate
            MaxTime = argument
        End Sub
        Private Function CalculateNextValue(ByVal value As Double) As Double
            Return value + (random.NextDouble() * 10.0 - 5.0)
        End Function
    End Class

    Public Structure ProcessItem
        Public Property DateAndTime() As Date
        Public Property Process1() As Double
        Public Property Process2() As Double
    End Structure

    Public Class DataCollection
        Inherits ObservableCollection(Of ProcessItem)

        Public Sub AddRange(ByVal items As IList(Of ProcessItem))
            For Each item As ProcessItem In items
                Me.Items.Add(item)
            Next item
            OnCollectionChanged(New NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, DirectCast(items, IList), Me.Items.Count - items.Count))
        End Sub
        Public Sub RemoveRangeAt(ByVal startingIndex As Integer, ByVal count As Integer)
            Dim removedItems = New List(Of ProcessItem)(count)
            For i As Integer = 0 To count - 1
                removedItems.Add(Items(startingIndex))
                Items.RemoveAt(startingIndex)
            Next i
            If count > 0 Then
                OnCollectionChanged(New NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, DirectCast(removedItems, IList), startingIndex))
            End If
        End Sub
    End Class
End Namespace
